File: android_native/app/src/main/java/com/ocai/app/DeepSeekService.java
package com.ocai.app;

import org.json.JSONArray;
import org.json.JSONObject;
import java.io.IOException;
import okhttp3.*;

public class DeepSeekService {

    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
    private static final OkHttpClient client = new OkHttpClient.Builder()
            .connectTimeout(15, java.util.concurrent.TimeUnit.SECONDS)
            .readTimeout(20, java.util.concurrent.TimeUnit.SECONDS)
            .build();

    public interface Callback {
        void onSuccess(String keywords);
        void onError(String fallback);  // fall back to raw input on failure
    }

    /**
     * Extract product search keywords from user input (async)
     */
    public static void extractKeywords(String userInput, Callback callback) {
        new Thread(() -> {
            try {
                String keywords = Config.USE_BACKEND_PROXY
                        ? callBackendProxy(userInput)
                        : callDeepSeekDirect(userInput);
                callback.onSuccess(keywords);
            } catch (Exception e) {
                callback.onError(userInput);  // fall back to raw input on any error
            }
        }).start();
    }

    // Direct DeepSeek call (development mode)
    private static String callDeepSeekDirect(String userInput) throws Exception {
        String system = "Ты помощник по покупкам. Извлеки ключевые слова для поиска товара из сообщения пользователя.\n"
                + "Верни ТОЛЬКО ключевые слова на русском языке, максимум 5 слов, без пояснений.\n"
                + "Пример: ввод \"хочу купить зимние кроссовки\" → вывод \"зимние кроссовки\"";

        JSONObject body = new JSONObject();
        body.put("model", Config.DEEPSEEK_MODEL);
        JSONArray messages = new JSONArray();
        messages.put(new JSONObject().put("role", "system").put("content", system));
        messages.put(new JSONObject().put("role", "user").put("content", userInput));
        body.put("messages", messages);
        body.put("max_tokens", 50);
        body.put("temperature", 0.3);

        Request request = new Request.Builder()
                .url(Config.DEEPSEEK_URL)
                .addHeader("Authorization", "Bearer " + Config.DEEPSEEK_API_KEY)
                .post(RequestBody.create(body.toString(), JSON))
                .build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) throw new IOException("HTTP " + response.code());
            JSONObject result = new JSONObject(response.body().string());
            return result.getJSONArray("choices")
                    .getJSONObject(0)
                    .getJSONObject("message")
                    .getString("content")
                    .trim();
        }
    }

    // Backend proxy call (production mode)
    private static String callBackendProxy(String userInput) throws Exception {
        JSONObject body = new JSONObject();
        body.put("input", userInput);

        Request request = new Request.Builder()
                .url(Config.BACKEND_URL + "/api/keywords")
                .post(RequestBody.create(body.toString(), JSON))
                .build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) throw new IOException("HTTP " + response.code());
            JSONObject result = new JSONObject(response.body().string());
            return result.getString("keywords");
        }
    }
}

File: lib/services/deepseek_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config.dart';

class DeepseekService {
  // 调用 DeepSeek 提取商品搜索关键词
  static Future<String> extractKeywords(String userInput) async {
    const systemPrompt = '''Ты помощник по покупкам на сайте ocai.ru (покупка товаров из Китая).
Пользователь описывает товар, который хочет купить.
Твоя задача: извлечь ключевые слова для поиска товара.

Правила:
- Верни ТОЛЬКО поисковые ключевые слова на русском языке
- Максимум 5 слов, самые важные
- Никаких пояснений, только слова
- Если пользователь уже написал конкретный товар — верни как есть

Примеры:
Ввод: "хочу купить удобные кроссовки для бега"
Вывод: кроссовки для бега

Ввод: "нужна детская куртка на зиму размер 110"
Вывод: детская куртка зимняя 110''';

    try {
      final response = await http.post(
        Uri.parse('$deepseekBaseUrl/chat/completions'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $deepseekApiKey',
        },
        body: jsonEncode({
          'model': deepseekModel,
          'messages': [
            {'role': 'system', 'content': systemPrompt},
            {'role': 'user', 'content': userInput},
          ],
          'max_tokens': 50,
          'temperature': 0.3,
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(utf8.decode(response.bodyBytes));
        final keywords = data['choices'][0]['message']['content'].toString().trim();
        return keywords;
      } else {
        // API 失败时直接用原始输入
        return userInput;
      }
    } catch (e) {
      // 网络异常时直接用原始输入
      return userInput;
    }
  }
}
